proposals/0529-th-quotes-as-patterns.rst

```diff
@@ -0,0 +1,513 @@
+
+#. This is not allowed::
+
+     f [d| _ = $(x); _ = $(y) |] = ...
+
+   because matching on lists of multiple declarations is left as future work.
+
+Optional: Fine-grained Quotation constraints
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+#. ``[| 1 + 1 |]`` would have type ``Monad m => m Exp`` instead of ``Quote m => m Exp``.
+
+#. ``[| \x -> x |]`` however would continue to have type ``Quote m => m Expr``.
+
+(Note, we could relax ``Monad`` to ``Applicative`` at an any point, but ``Monad`` is the current superclas of ``Quote``.)
+
+Effect and Interactions
+-----------------------
+
+``TemplateHaskell`` will not imply ``TemplateHaskellQuotesAsPatterns``, though it does imply ``TemplateHaskellQuotes``.
+
+The banned binding constructs are precisely those which would need ``newName`` in expression position.
+
+Optional Proposed Change: Fine-grained Quotation constraints
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The optional proposed relaxation of the expression position rules is supposed to make these pattern-position restrictions more familiar to the programmer.
+Specifically, by distinguishing the same subset of quotes in two ways (they're the only ones allowed in pattern position, they get a more general type in expresssion position), we give programmers two different ways to learn the difference between them and quotes in general.
+
+If we go with the alternative of supporting local variables, however, this extra motivation dries up.
+The cases where ``newName`` is needed in expression position will not no longer correspond to anything obvious in pattern position in this alternative.
+The originally motivation from `Proposal #246`_ of more flexible typing for flexibility's sake still stands, however.
+
+Costs and Drawbacks
+-------------------
+
+Action asymmetry
+~~~~~~~~~~~~~~~~
+
+The lack of symmetry where expression create actions but patterns only bind plain AST values is annoying.
+But the fixes for this might be too radical?
+
+Alternatives
+------------
+
+Pattern match on actions(!)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+I hypothesize that we could do better than the proposed actions vs no action asymmetry by meditating on the ways pattern matching relates to optics.
+For example::
+
+  [| a + $(x) |]
+
+could match any ``t Exp`` where ``t`` is ``Traversable`` and bind ``x :: t Expr`` with this desugaring::
+
+  f [| a + $(x) |] = ...
+
+  =>
+
+  f (traverse __inner -> Just x) = ...
+
+  __inner (AppE (VarE __a) x) | __a == 'a = Just x
+  __inner _                               = Nothing
+
+This behavior seems overwrought, as we are doing the as-proposed behavior *plus* an additional traversal.
+But this matches the fact that expression-position quotes are do what idiom brackets do (implicit ``Applicative``) in addition to base quoting.
+
+Support local variables
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Quotes that bind local variables do in fact have an interpretation as non-linear patterns::
```

Comment:
In macro systems like Racket's and Lean's, you cannot distinguish between bindings and uses in quotations because scoping is discovered only during expansion. So "quotes that bind local variables" is not even something we can detect in those systems.

Reply:
I presume "discovered only during expansion" is an artifact of the fact that  Racket-style macros will not get a name-resolved TH AST as input (where we know the binding structure but limit ourselves to macro syntax expressible in Haskell) but rather just a concrete syntax tree. Perhaps let me rephrase in terms of the example in https://school.racket-lang.org/2019/plan/tue-mor-lecture.html:

```
(define-syntax (robust-for/list5 stx)
  (syntax-parse stx
    [(_ ([elem-name:id seq]) computation)
     #'(map (λ (elem-name) computation) seq)]))
```

This macro will expand `(robust-for/list5 ([x (list 1 2)]) (add1 x))` to `(map (λ (x) (add1 x)) (list 1 2))`, matching `[elem-name :-> x, seq :-> (list 1 2)]`. Note that `elem-name` is ultimately spliced into a binding occurrence and that `computation` may refer to the instantiation of `elem-name`, thus its *syntax class* is specified to be `id`, which is what `elem-name:id` does. (This is somewhat like pushing an expected type for the AST node to match inside the pattern, I guess, where "type" in Racket really means a user-definable non-terminal of the extensible syntax. See 5.6 in the link above.) But the *pattern* `(_ ([elem-name:id seq]) computation)` does not know about binding structure at all; rather the macro equips the pattern with a binding structure post-hoc when it generates the template `(map (λ (elem-name) computation) seq)]))` for the pattern, thus intentionally capturing occurrences of `elem-name` in `computation`.

Reply:
That's exactly right. It's the same reason I mentioned in the ZuriHac discussion why our macros are untyped (on the level of the quoted language) but core-level transformations can optionally be typed (which is done by https://github.com/gebner/quote4 btw).